sunpy/io/special/asdf/converters/generic_map.py

```diff
@@ -12,25 +11,36 @@ class GenericMapConverter(Converter):
     @property
     def types(self):
         from sunpy.map.mapbase import GenericMap
-        return list(GenericMap._registry.keys())
+        return [GenericMap] + list(GenericMap._registry.keys())
+
+    def select_tag(self, obj, tags, ctx):
+        # Sort the tags in reverse alphabetical order and pick the first (i.e.
+        # the one with the highest version). This assumes that all the tags for
+        # this converter are named the same other than the version number.
+        tags = list(sorted(tags, reverse=True))
+        return tags[0]
 
     def from_yaml_tree(self, node, tag, ctx):
+        import astropy.units as u
+
         import sunpy.map
 
         # Use the factory here to get the correct subclass back
         out_map = sunpy.map.Map(np.asanyarray(node["data"]), node["meta"])
-        out_map = out_map.shift(*node["shift"])
-
         out_map.mask = node.get("mask")
         out_map.uncertainty = node.get("uncertainty")
         out_map._unit = node.get("unit")
+
+        # Old files can have a shift that needs applying
```

Comment:
The existing shift method set the `shift` value *and* updated the Map `meta`. So was `from_yaml_tree` shifting it twice? Or maybe the `shift` was `0º` in this case?

Reply:
`GenericMap.shift` does:

```py
new_meta['crval1'] = ((self._reference_longitude + axis1).to(self.spatial_units[0])).value
new_meta['crval2'] = ((self._reference_latitude + axis2).to(self.spatial_units[1])).value
```

and

```py
new_map._shift = SpatialPair(self.shifted_value[0] + axis1,
                             self.shifted_value[1] + axis2)
```

If you then save to ASDF and run `sunpy.io.special.asdf.converters.generic_map.from_yaml_tree` it does:

```py
out_map = sunpy.map.Map(np.asanyarray(node["data"]), node["meta"])
```

which already includes the shift (as `crval1` and `crval2` were already updated by `GenericMap.shift`), however it also does:

```py
out_map = out_map.shift_reference_coord(*node['shift'])
```

Which shifts it again?

Reply:
So both these things don't happen together.

We no longer save the shift explicitly (see the `to_yaml_tree` method), it is just folded in to `meta`. What this code supports is loading files which were written with sunpy <4.0 (genericmap tag version 1.0.0), where the shift is explicitly saved in the file and we need to apply it on load.